lib/bamboo/api_error.ex

```diff
@@ -0,0 +1,34 @@
+defmodule Bamboo.ApiError do
+  @moduledoc """
+  Error used to represent a problem when sending emails through an external email service API.
+  """
+
+  defexception [:message]
+
+  def raise_api_error(message), do: raise(__MODULE__, message: message)
+  def raise_api_error(adapter, response, params, extra_message \\ "") when is_atom(adapter) do
+    message = """
+    There was a problem sending the email through the #{adapter_name(adapter)} API.
```

Comment:
I think that's a great solution.

You may also consider a map or keyword list so that the arguments can be named which would make it more clear.